src/number_theory/modular.lean

```diff
@@ -363,6 +393,269 @@ begin
       simp [T', sub_eq_add_neg] } }
 end
 
+
+/-- Crucial lemma showing that if `c≠0`, then `3/4 < 4/(3c^4)` -/
+lemma ineq_1 (z : ℍ) (g: SL(2,ℤ)) (hz : z ∈ 𝒟ᵒ) (hg: g • z ∈ 𝒟ᵒ) (c_ne_z : g 1 0 ≠ 0) :
+  (3 : ℝ)/4 < 4/ (3* (g 1 0)^4) :=
+begin
+  have z_im := z.im_ne_zero,
+  have c_4_pos : (0 : ℝ) < (g 1 0)^4,
+    exact_mod_cast (by simp: even 4).pow_pos c_ne_z ,
+  /- Any point `w∈𝒟ᵒ` has imaginary part at least `sqrt (3/4)` -/
+  have ImGeInD : ∀ (w : ℍ), w ∈ 𝒟ᵒ → 3/4 < (w.im)^2,
+  { intros w hw,
+    have : 1 < w.re * w.re + w.im * w.im := by simpa [complex.norm_sq_apply] using hw.1,
+    have := hw.2,
+    cases abs_cases w.re; nlinarith, },
+  /- The next argument is simply that `c^2 y^2 ≤ |c z + d|^2`. -/
+  have czPdGecy : (g 1 0 : ℝ)^2 * (z.im)^2 ≤ norm_sq (denom g z) :=
+    calc
+    (g 1 0 : ℝ)^2 * (z.im)^2 ≤ (g 1 0 : ℝ)^2 * (z.im)^2 + (g 1 0 * z.re + g 1 1)^2 : by nlinarith
+    ... = norm_sq (denom g z) : by simp [norm_sq]; ring,
+  have zIm : (3 : ℝ) / 4 < (z.im)^2 := ImGeInD _ hz,
+  /- This is the main calculation:
+  `sqrt 3 / 2 < Im(g•z) = Im(z)/|cz+d|^2 ≤ y/(c^2 y^2) < 2/(c^2 sqrt 3)`
+  -/
+  calc
+  (3 : ℝ) / 4 < ((g • z).im) ^ 2 : ImGeInD _ hg
+  ... = (z.im) ^ 2 / (norm_sq (denom g z)) ^ 2 : _
+  ... ≤ (1 : ℝ) / ((g 1 0) ^ 4 * (z.im) ^ 2) : _
+  ... < (4 : ℝ) / (3 * (g 1 0) ^ 4) : _,
+  { convert congr_arg (λ (x:ℝ), x ^ 2) (im_smul_eq_div_norm_sq g z) using 1,
+    exact (div_pow _ _ 2).symm, },
+  { rw div_le_div_iff,
+    convert pow_le_pow_of_le_left _ czPdGecy 2 using 1;
+    ring_nf,
+    { nlinarith, },
+    { exact pow_two_pos_of_ne_zero _ (norm_sq_denom_ne_zero g z), },
+    { nlinarith, }, },
+  { rw div_lt_div_iff,
+    repeat {nlinarith}, },
+end
+
+
+/-- Knowing that `3/4<4/(3c^4)` from `ineq_1`, and `c≠0`, we conclude that `c=1` or `c=-1`. -/
+lemma ineq_2 (c : ℤ) (hc₁ : (3 : ℝ)/4 < 4/ (3* c^4)) (hc₂ : c ≠ 0) : c = 1 ∨ c = -1 :=
+begin
+  rcases le_or_gt (|c|) 1 with h | (h : 2 ≤ |c|),
+  { -- case |c| ≤ 1
+    obtain ⟨h1c, hc1⟩ : -1 ≤ c ∧ c ≤ 1 := abs_le.mp h,
+    interval_cases c; tauto },
+  { -- case 2 ≤ |c|
+    exfalso,
+    have : 2^4 ≤ c^4,
+    { refine pow_four_le_pow_four _,
+      convert h using 1, },
+    have : (2:ℝ)^4 ≤ c^4,
+    { norm_cast,
+      convert this using 1, },
+    have := (div_lt_div_iff _ _).mp hc₁,
+    repeat {linarith}, },
+end
+
+/-- Definition: `T_pow` is the matrix `T` raised to the power `n:ℤ`. -/
```

Comment:
Can you add a lemma that states this is actually true?

Reply:
Alternatively, perhaps `T_pow` shouldn't exist, and you should just have a lemma that expands `T ^ n` into this form.